nemo/core/neural_factory.py

```diff
@@ -84,6 +192,44 @@ class DeviceType(Enum):
     AllGpu = 3
 
 
+class TrainingState:
+    def __init__(self):
+        tensor_naming_registery = AppState().tensor_names
+        self.tensor_dict = {}.fromkeys(tensor_naming_registery.unique_names, None)
+
+    def tensor_list(self):
+        return self.tensor_dict.keys()
+
+    def clear_dict(self):
+        for name in self.tensor_dict:
+            self.tensor_dict[name] = None
+
+    def set_tensor(self, tensor, value):
+        self.tensor_dict[tensor.unique_name] = value
+
+    def check_tensor_cached(self, unique_name):
+        if self.tensor_dict[unique_name] is None:
+            return False
+        return True
+
+    def get_tensor(self, name):
+        unique_name = AppState().tensor_names[name]
+        return self.tensor_dict[unique_name]
+
+    def get_and_compute_tensor(self, name, action):
+        unique_name = AppState().tensor_names[name]
+        tensor_value = self.tensor_dict[unique_name]
+        if tensor_value is None:
+            nmtensor = AppState().tensor_names._nmtensor_uniname_dict[unique_name]
+            callchain = topological_sort_from_leaves([nmtensor], cached_training_state=self)
+            # print(callchain)
+            callchain.insert(0, ())
+            action.nm_graph_forward_pass(callchain, self.tensor_dict)
+            # print(self.tensor_dict[unique_name])
+            tensor_value = self.tensor_dict[unique_name]
+        return tensor_value
+
+
 class Actions(ABC):
```

Comment:
Discussed with @blisc : move Actions + TrainingState + graph traversing-related code to a separate file

Reply:
Moved Actions, TrainingState and topological_sort_from_leaves to a new file called nemo/core/actions.py